src/Psl/Shell/execute.php

```diff
@@ -0,0 +1,89 @@
+ * @param string                $working_directory  The initial working directory for the command.
+ *                                                  This must be an absolute directory path, or null
+ *                                                  if you want to use the default value (the working dir of the
+ *                                                  current directory )
+ * @param array<string, string> $environment        An array with the environment variables for the command that
+ *                                                  will be run.
+ * @param bool                  $escape_arguments   If set to true ( default ), all $arguments will be escaped using
+ *                                                  `Shell\escape_argument`.
+ *
+ * @psalm-taint-sink shell $command
+ *
+ * @throws Exception\FailedExecutionException   In case the command resulted in an exit code other than 0.
+ * @throws Exception\PossibleAttackException    In case the command being run is suspicious ( e.g: contains NULL byte ).
+ * @throws Exception\RuntimeException           In case $working_directory doesn't exist, or unable to create a new
+ *                                              process.
+ */
+function execute(
+    string $command,
+    array $arguments = [],
+    ?string $working_directory = null,
+    array $environment = [],
+    bool $escape_arguments = true
+): string {
+    if ($escape_arguments) {
+        $arguments = Vec\map($arguments, static fn(string $argument): string => escape_argument($argument));
+    }
+
+    $commandline = Str\join([$command, ...$arguments], ' ');
+
+    /** @psalm-suppress MissingThrowsDocblock - safe ( $offset is within-of-bounds ) */
+    if (Str\contains($commandline, "\0")) {
+        throw new Exception\PossibleAttackException('NULL byte detected.');
+    }
+
+    $descriptor = [
+        1 => ['pipe', 'w'],
+        2 => ['pipe', 'w'],
+    ];
+
+    $environment = Dict\merge(Env\get_vars(), $environment);
+    $working_directory = $working_directory ?? Env\current_dir();
+    if (!is_dir($working_directory)) {
+        throw new Exception\RuntimeException('$working_directory does not exist.');
+    }
+
+    $process = proc_open($commandline, $descriptor, $pipes, $working_directory, $environment);
+    // @codeCoverageIgnoreStart
+    // not sure how to replicate this, but it can happen \_o.o_/
+    if (!is_resource($process)) {
+        throw new Exception\RuntimeException('Failed to open a new process.');
+    }
+    // @codeCoverageIgnoreEnd
+
+    $stdout_content = stream_get_contents($pipes[1]);
+    $stderr_content = stream_get_contents($pipes[2]);
+
+    fclose($pipes[1]);
+    fclose($pipes[2]);
+
+    $code = proc_close($process);
+    if ($code !== 0) {
+        throw new Exception\FailedExecutionException($commandline, $stdout_content, $stderr_content, $code);
+    }
+
+    return $stdout_content;
```

Comment:
stdout content could also be usefull.

Reply:
you mean stderr? 

yes, but not in most case, `Shell\execute` is a replacement for `exec` and `shell_exec`, not `proc_*` ( that will be replaced probably by `Psl\Process` once we have #127 ), and these functions don't return the content in stderr, instead it's redirected to PHPs own stderr, but that's something i don't think we should do.

Reply:
Yes stderr indeed.

What addition would the Psl process component have on top of Shell\Execute then? Why not go for 1 way of interacting with processes? 

If you only return stdout, you loose stderr if the command finished correctly.
So in case of the `composer run` example, you miss information.

Maybe return a chronological stderr + stdout string as a result instead?

Reply:
> If you only return stdout, you loose stderr if the command finished correctly.
> So in case of the composer run example, you miss information.

that's the same with `exec` and `shell_exec`, which `Shell\execute` is meant to replace.

the advantage of using `Shell\execute` instead of `exec`/`shell_exec`:

1. in case of an error, a `Shell\Exception\FailedExecutionException` is thrown instead of an error+stderr being redirected to phps stderr.
2. you can change working directory
3. you can add additional environment variables
4. it escapes arguments for you by default

if you want to have an interactive process + always get your hand on the stderr output of the underlying process, you should probably use `proc_*` or `symfony/process` directly for now, until we have IO handles, in which case we can introduce `Psl\Process` component that would be more powerful, and would be able to replace `proc_*` and `symfony/process` completely.

we want to keep this simple

Reply:
My expectations of this component might have been different then yours.
If you are sure about the API and implementation, then I don't see any reason not to merge it.

Reply:
you should wait for `Psl\Process` 😉 already on the work